### PR TITLE
feat: KnowledgeGraph builder 

### PR DESCRIPTION
**Pull request by debrupf2946**
The GraphRag folder contains codes that allow users to build a Knowledge Graph from documents.

### CODE REVIEW

**Comment by kmehant** [comment]
@debrupf2946 The README is poorly written with scant attention to grammar and formatting please update it.

**Review by kmehant — CHANGES_REQUESTED** [review]
Rename the folder GraphRag to snake case and give it informative naming such graph_rag_using_llamaindex etc

File: GraphRag/README.MD [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+1.Now simply build the KG using the [documents](###DataPreperation)  
+2. call ***build_graph(documents)*** this will creat an index  
+3. It wil also save Graph_visualization.html,which can be opened in browser to visualise the KowledgeGraph built
+```python
+index=build_graph(documents)
+```
+4. Save the ***index*** as pickle file
+```python
+save_index(index)
+```
```
Comment (kmehant) [inline GraphRag/README.MD:48]: Give one end to end usage example with proper steps written with screenshots or GIFs wherever necessary.

File: GraphRag/knowledgeGraph.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    """
+    serializes the index object,so that it can be loaded and used later
+    Args:
+        index: Grpah-Index object
+
+    Returns:
+        saves pickle file of the Grpah-Index
+    """
+    os.makedirs("results", exist_ok=True)
+    with open('results/graphIndex', 'wb') as f:
```
Comment (kmehant) [inline GraphRag/knowledgeGraph.py:52]: Take path as part of the input rather hardcoding it.

File: GraphRag/knowledgeGraph.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        embed_model =HuggingFaceEmbedding(model_name=embeddings),
+        storage_context=storage_context,
+    )
+    print("KG built succesfully!")
+    os.makedirs("results", exist_ok=True)
+    g = index.get_networkx_graph()
+    net = Network(notebook=True, cdn_resources="in_line", directed=True)
+    net.from_nx(g)
+    net.show("Graph_visualization.html")
+    return index
```
Comment (kmehant) [inline GraphRag/knowledgeGraph.py:40]: In the entirety of the code base please add proper exception handling.

File: GraphRag/knowledgeGraph.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+from llama_index.core import StorageContext
```
Comment (kmehant) [inline GraphRag/knowledgeGraph.py:1]: Explain purpose of each file as a docstring at the top of each python file.

File: GraphRag/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+llama-index-embeddings-huggingface
+llama-index-llms-ollama
+llama-index
+pyvis
+tree-sitter==0.21.3
+tree-sitter-languages
```
Comment (kmehant) [inline GraphRag/requirements.txt:6]: Where are we using tree-sitter package in the code?

File: GraphRag/tools.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+from llama_index.llms.ollama import Ollama
+from llama_index.core import SimpleDirectoryReader
+from llama_index.core.node_parser import CodeSplitter
+from llama_index.core import Document
+from llama_index.core import Settings
+
+
+
+
+def initialize_llm(base_url: str ="http://localhost:11434",model: str ="llama3",chunk_size: int = 512):
```
Comment (kmehant) [inline GraphRag/tools.py:10]: Format your code using black tool. This review comment is meant for all the code files.

**Review by kmehant — CHANGES_REQUESTED** [review]
Require more changes.

**Review by kmehant — CHANGES_REQUESTED** [review]
@debrupf2946 

Update your PR with more information description on the tooling and approach you have taken in this PR.  Add screenshots of examples that you have tried with this contribution.

**Comment by debrupf2946** [comment]
> @debrupf2946 The README is poorly written with scant attention to grammar and formatting please update it.
Thank you for the review
I have updated the README.MD

File: GraphRag/knowledgeGraph.py (hunk 3 above)
Comment (debrupf2946) [reply GraphRag/knowledgeGraph.py:40]: I had that in my mind, I will be doing it also I will be using PyDantic to TypeCheck

File: GraphRag/requirements.txt (hunk 5 above)
Comment (debrupf2946) [reply GraphRag/requirements.txt:6]: When we are using a code splitter it uses the same internally

Comment (kmehant) [reply GraphRag/requirements.txt:6]: Then the code splitter should be installing it internally you should not be needed to provide it as a dependency for your own module.

File: GraphRag/knowledgeGraph.py (hunk 3 above)
Comment (kmehant) [reply GraphRag/knowledgeGraph.py:40]: Please update the PR with the same.

**Review by kmehant — CHANGES_REQUESTED** [review]
update on review

File: GraphRag/README.MD (hunk 1 above)
Comment (debrupf2946) [reply GraphRag/README.MD]: I have created an example folder that contains an collab file showing detailed implementation

File: GraphRag/requirements.txt (hunk 5 above)
Comment (debrupf2946) [reply GraphRag/requirements.txt:6]: when I was trying in collab file there was a bug appearing which got solved only after installing `tree-sitter==0.21.3`
`tree-sitter-languages` they weren't installed internally that's why I added them to make code bug free

File: GraphRag/tools.py (hunk 6 above)
Comment (debrupf2946) [reply GraphRag/tools.py]: Done! codes are formatted with black

File: GraphRag/knowledgeGraph.py (hunk 4 above)
Comment (debrupf2946) [reply GraphRag/knowledgeGraph.py:5]: added docstring for each file

File: GraphRag/knowledgeGraph.py (hunk 2 above)
Comment (debrupf2946) [reply GraphRag/knowledgeGraph.py]: I updated the code and removed hardcoding

File: GraphRag/README.MD [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+```python
+index = build_graph(documents)
+```
+
+4. Save the `index` as a pickle file.
+
+```python
+save_index(index)
+```
```
Comment (kmehant) [inline GraphRag/README.MD:64]: refer example notebook here.

File: GraphRag/requirements.txt (hunk 5 above)
Comment (kmehant) [reply GraphRag/requirements.txt:6]: Okay

**Review by kmehant — CHANGES_REQUESTED** [review]
move all this to `graph_rag/graph_builder` folder instead of `GraphRag` look at my other comments that I have added now. Thanks

**Comment by debrupf2946** [comment]
> graph_builder

Refactored codes to graph_rag/graph_builder

File: GraphRag/README.MD (hunk 7 above)
Comment (kmehant) [reply GraphRag/README.MD]: @debrupf2946  ^

Comment (debrupf2946) [reply GraphRag/README.MD]: done!

**Review by kmehant — APPROVED** [review]
LGTM

**Comment by kmehant** [comment]
@debrupf2946 update the PR with signed commit and we are good to go.

**Pull request merged**